File: Data Structures & Algorithms/coin-change/submission-0.cpp
class Solution
 {
public:
int f(int i,vector<int>& coins, int amount, vector<vector<int>>&dp)
{
    if(amount==0)
    return 0;
    if(i==coins.size() || amount<0)
    {
        return INT_MAX;
    }
    if(dp[i][amount]!=-1)
    {
        return dp[i][amount];
    }
    int take=INT_MAX;
    int nottake=INT_MAX;
    if(coins[i]<=amount)
    {
        int res=f(i,coins,amount-coins[i],dp);
        if(res!=INT_MAX)
        {
            take=1+res;
        }
    }
    nottake=f(i+1,coins,amount,dp);
    return  dp[i][amount]=min(take,nottake);
}
    int coinChange(vector<int>& coins, int amount) 
    {
        int n=coins.size();

        vector<vector<int>>dp(n,vector<int>(amount+1,-1));
                int ans=f(0,coins,amount,dp);
        return ans==INT_MAX?-1:ans;
        
    }
};

File: Data Structures & Algorithms/coin-change-ii/submission-0.java
class Solution
 {
    int f(int i,int amount, int[] coins,   int [][]dp)
    {
        if(i==coins.length-1)
        {
            return amount%coins[i]==0?1:0;
        }
        int take=0;
        int nottake=0;
        if(dp[i][amount]!=-1)
        return dp[i][amount];
        if(coins[i]<=amount)
        {
            take=f(i,amount-coins[i],coins,dp);
        }
        nottake=f(i+1,amount,coins,dp);
        return dp[i][amount]=take+nottake;
    }
    public int change(int amount, int[] coins)
     {
        int n=coins.length;
        int [][]dp=new int[n][amount+1];
        for(int[] row:dp)
        {
            Arrays.fill(row,-1);
        }
        return f(0,amount,coins,dp);
    }
}